File: ios/App/App/NativeRouteDispatcher.swift
import Foundation

enum NativeRouteDispatcher {
    static func parseTrackOrderId(_ pathOrUrl: String) -> String? {
        guard let pathname = extractPathname(pathOrUrl) else { return nil }
        let pattern = #"^/orders/([^/]+)/track/?$"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return nil }
        let range = NSRange(pathname.startIndex..<pathname.endIndex, in: pathname)
        guard let match = regex.firstMatch(in: pathname, range: range),
              match.numberOfRanges > 1,
              let idRange = Range(match.range(at: 1), in: pathname)
        else { return nil }
        return String(pathname[idRange])
    }

    static func shouldOpenNativeTrack(hostEnabled: Bool, trackEnabled: Bool, inCohort: Bool) -> Bool {
        hostEnabled && trackEnabled && inCohort
    }

    private static func extractPathname(_ pathOrUrl: String) -> String? {
        let raw = pathOrUrl.trimmingCharacters(in: .whitespacesAndNewlines)
        if raw.hasPrefix("/") {
            return raw.split(separator: "?").first.map(String.init)
        }
        guard let url = URL(string: raw) else { return nil }
        let host = url.host ?? ""
        let path = url.path
        if host == "app" || host.isEmpty {
            return path
        }
        return "/\(host)\(path)"
    }
}

File: android/app/src/main/java/com/bhojanos/orderbhojan/nativehost/NativeRouteDispatcher.kt
package com.bhojanos.orderbhojan.nativehost

/**
 * Route ID `track` → `/orders/:orderId/track`
 */
object NativeRouteDispatcher {
    private val TRACK_PATH = Regex("""^/orders/([^/]+)/track/?$""")

    fun parseTrackOrderId(pathOrUrl: String): String? {
        val pathname = extractPathname(pathOrUrl) ?: return null
        return TRACK_PATH.matchEntire(pathname)?.groupValues?.getOrNull(1)
    }

    fun shouldOpenNativeTrack(
        hostEnabled: Boolean,
        trackEnabled: Boolean,
        inCohort: Boolean,
    ): Boolean = hostEnabled && trackEnabled && inCohort

    private fun extractPathname(pathOrUrl: String): String? {
        val raw = pathOrUrl.trim()
        if (raw.startsWith("/")) {
            return raw.substringBefore('?').substringBefore('#')
        }
        return try {
            val uri = android.net.Uri.parse(raw)
            val host = uri.host.orEmpty()
            val path = uri.path.orEmpty()
            when {
                host == "app" || host.isEmpty() -> path
                else -> "/$host$path"
            }.substringBefore('?')
        } catch (_: Exception) {
            null
        }
    }
}
